### PR TITLE
NMS.FT.MobileAcademyControllerIT

### PR DESCRIPTION
**Pull request by AnkitGaurNms**
Implemented FT 564
JIRA :https://applab.atlassian.net/browse/NMS-250

### CODE REVIEW

File: testing/src/main/java/org/motechproject/nms/testing/web/TestingController.java [hunk 1]
```diff
@@ -29,4 +33,10 @@ public String sendEvent(@RequestParam(required = true) String subject) {
         return String.format("Sent MOTECH event: %s with empty params: %s", subject, eventParams.toString());
     }
 
+    @RequestMapping(value = "/sendSMS200", method = RequestMethod.POST, headers = { "Content-type=application/json" })
+    @ResponseStatus(HttpStatus.ACCEPTED)
```
Comment (koshalt) [inline testing/src/main/java/org/motechproject/nms/testing/web/TestingController.java]: This looks wrong to me. Accepted is 201 and OK is 200?

Comment (AnkitGaurNms) [inline testing/src/main/java/org/motechproject/nms/testing/web/TestingController.java]: @koshalt this method return response status code as 202 (Accepted) and response entity as "OK".

Comment (koshalt) [inline testing/src/main/java/org/motechproject/nms/testing/web/TestingController.java]: Yes, I understand that but you cannot call an API sendSMS200 and send status code 202 for it. That's bad practice. Please change.

File: testing/src/test/java/org/motechproject/nms/testing/it/api/MobileAcademyControllerBundleIT.java [hunk 2]
```diff
@@ -1003,5 +1128,210 @@ public void verifyFT532() throws IOException, InterruptedException {
         assertTrue("{\"bookmark\":\"Chapter01_Lesson01\",\"scoresByChapter\":{\"1\":2}}"
                 .equals(EntityUtils.toString(response.getEntity())));
     }
+
+    /**
+     * To check delivery notification API is successful when optional parameter
+     * call data is missing.
+     */
+    // TODO https://applab.atlassian.net/browse/NMS-250
+    @Ignore
```
Comment (koshalt) [inline testing/src/test/java/org/motechproject/nms/testing/it/api/MobileAcademyControllerBundleIT.java]: There is a fix coming in for NMS-250 but this test needs a different design. Sms status updates are handled asynchronously while this test tries to test it synchronously, causing the completion record to be read before the event handler code updates the status.

You should either sleep here or to be even more reliable, test the different layers (api vs update) differently.

**Comment by larubbio** [comment]
I'm a little concerned that the builds have been failing with this error message:

```
21:46:58 [WARNING] failed SelectChannelConnector@0.0.0.0:8181: java.net.BindException: Address already in use
21:46:58 [WARNING] failed Server@69c54bfb: java.net.BindException: Address already in use
21:46:58 [ERROR] Exception while initializing Jetty.
21:46:58 java.net.BindException: Address already in use
```

Around the same time we introduces tests that have the server issuing an http request to itself.  I'm not sure it is related but it might be.  

Currently it seems the test updates the server config so the url for sending an SMS is the server itself.  The test then issues an HTTP request to the server that causes the server to issue another http request to itself.  (I'm not sure how many threads we run while running these tests).

**Comment by larubbio** [comment]
ok to test

**Comment by srivastavaAbhi** [comment]
ok to test

**Comment by srivastavaAbhi** [comment]
test this please

**Comment by koshalt** [comment]
test this please

**Comment by koshalt** [comment]
test this please

File: testing/src/test/java/org/motechproject/nms/testing/it/imi/ImiSMSNotificationServiceBundleIT.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+                .getProperty(MAX_NOTIFICATION_RETRY_COUNT);
+    }
+
+    @After
+    public void restore() {
+        settingsFacade.setProperty(SMS_NOTIFICATION_URL, oldEndpoint);
+        settingsFacade.setProperty(MAX_NOTIFICATION_RETRY_COUNT, oldRetryCount);
+    }
+
+    @Test
```
Comment (koshalt) [inline testing/src/test/java/org/motechproject/nms/testing/it/imi/ImiSMSNotificationServiceBundleIT.java]: @srivastavaAbhi Why is this related to NIP-47? Also, this needs a test description for what it does.

**Comment by frankhuster** [comment]
@AnkitGaurNms, @srivastavaAbhi any reason nothing's happening with this PR?

File: testing/src/test/java/org/motechproject/nms/testing/it/imi/ImiSMSNotificationServiceBundleIT.java (hunk 3 above)
Comment (srivastavaAbhi) [inline testing/src/test/java/org/motechproject/nms/testing/it/imi/ImiSMSNotificationServiceBundleIT.java]: @koshalt  : We are trying to check the code behavior upon receiving 202 from IMI. This is the only relation of this test case with #434 .
I will add some description as well.

**Comment by srivastavaAbhi** [comment]
test this please

**Comment by srivastavaAbhi** [comment]
@koshalt : There is some error in jenkins setup because of which build is failed.
Could you help me out.

**Comment by srivastavaAbhi** [comment]
test this please

**Comment by frankhuster** [comment]
A month without action on this, I'll close it.
